fms/models/gpt_bigcode.py

```diff
@@ -473,7 +474,94 @@ def _hf_sd_to_fms_sd(hf_sd: Mapping) -> Mapping:
     return new_sd
 
 
+def _gptq_fused_sd_to_fms_unfused_sd(hf_sd: Mapping) -> Mapping:
+    """Fused-checkpoint-to-unfused-GPTQ-model adapter.
+    Works for GPTQ models instantiated as unfused using unfused_strategy="pre"
+    """
+
+    replacements = [
+        ("lm_head.weight", "head.weight"),
+        (r"^transformer.wte.weight", "base_model.embedding.weight"),
+        (r"^transformer.wpe.weight", "base_model.position_embedding.weight"),
+        (r"^transformer.ln_f", "base_model.dec_norm"),
+        (r"^transformer.h", "base_model.layers"),
+        # need to do kqv manually
+        (r"attn\.c_proj", "attn.dense"),
+        (r"mlp\.c_fc", "ff_sub_layer.w1"),
+        (r"mlp\.c_proj", "ff_sub_layer.w2"),
+        (r"ln_1", "ln"),
+        (r"ln_2", "ff_ln"),
+    ]
+
+    pack_ratio = 8  # assume 4 bits => packing/compression ratio = 32 / 4
+
+    # derive out_feature from g_idx (instead of passing it to the adapter)
+    # TODO: improve this, it may be fragile...
+    for hf_name, param in hf_sd.items():
+        if "attn.c_attn.g_idx" in hf_name:
+            query_out_feat = param.size(0)
```

Comment:
could this not be done inside the below loop when `"attn.c_attn" in hf_name`

Reply:
the adapter receives a subset of the state dict (partial_sd), which typically corresponds to all parameters from a single transformer layer. 

These parameters are processed in random order by the main loop (line 506) and there is no guarantee that `attn.c_attn.g_idx` (from which I extract `query_out_feat`) is processed before `qweight`, `qzeros`, and `scales` (which make use of `query_out_feat`)